gemd/entity/base_entity.py

```diff
@@ -105,14 +106,104 @@ def to_link(self,
+        """
+        if cache is None:
+            cache = {}
+        cache_key = frozenset((id(this), id(that)))
+        if cache_key in cache:
+            return cache[cache_key]
+        cache[cache_key] = None  # Mark as in progress
+
+        this_dict = this._dict_for_compare()
+        that_dict = that._dict_for_compare()
+        if this_dict.keys() != that_dict.keys():
+            cache[cache_key] = False  # Mark as failed
+            return False
+
+        # Check UIDs for an easy potential win
+        this_uids = this_dict.pop("uids")
+        that_uids = that_dict.pop("uids")
+        if this_uids != that_uids:
+            cache[cache_key] = False  # Mark as failed
+            return False
+
+        # Crawl rest of values to verify they align
+        for key in this_dict:
+            this_value = this_dict[key]
+            that_value = that_dict[key]
+            if isinstance(this_value, BaseEntity) and isinstance(that_value, BaseEntity):
+                if BaseEntity._cached_equals(this_value, that_value, cache=cache) is False:
+                    cache[cache_key] = False  # Mark as failed
+                    return False
+            elif isinstance(this_value, Collection) and isinstance(that_value, Collection) \
+                    and not isinstance(this_value, str) and not isinstance(that_value, str):
+                # Necessary to maintain context for recursive parts of the structure
+                this_list = list(this_value)
+                that_list = list(that_value)
+                if len(this_list) != len(that_list):
+                    # If an object was flattened, certain lists may be empty
+                    if len(this_list) == 0:
+                        if len(this_uids) > 0:  # But if it was flattened, uids will save us
+                            continue
+                    if len(that_list) == 0:
+                        if len(that_uids) > 0:  # But if it was flattened, uids will save us
+                            continue
+                    cache[cache_key] = False  # Mark as failed
+                    return False
+
+                # Finally crawl and compare
+                for x in this_list:
+                    found = False
+                    i_found = None
+                    if isinstance(x, BaseEntity):
+                        for i_found, y in enumerate(that_list):
+                            # Unless something really broke, y is a BaseEntity
+                            result = BaseEntity._cached_equals(x, y, cache=cache)
+                            if result is True:
+                                found = True
+                                break
+                            elif result is None:
+                                # Don't know yet; pass as False will appear elsewhere
+                                found = None
+                    else:
+                        found = x in that_list
```

Comment:
why do we not set `i_found` in this branch?

Reply:
Funnily enough, that's how I wrote it originally.  And then python explained that `in` is an implicit equality comparison, resulting in infinite descent.